### ISSUE TITLE
Return as exception all mandatory fields

### ISSUE THREAD

**Issue by bettinz** [osteel/openapi-httpfoundation-testing#27]
## Detailed description

Right now I have the `id` and `name` fields mandatory and `description` optional.
If I don't send `id` and `name` the exception return only `id` as missing. Can I show all missing fields that are mandatory?

Thanks

**Comment by osteel** [comment]
Hi,

Can I see the relevant section of your OpenAPI definition please?

**Comment by bettinz** [comment]
Yes, it’s already attached in the previous message (the second block) :-)

**Comment by osteel** [comment]
Oh sorry, overlooked it and thought it was an example payload.

Right so I think the issue comes from `allOf`, that doesn't quite behave the way you would expect. You'd think it would just merge the definitions, but it doesn't quite work like that (although I can't remember what it does off the top of my head – been a while since I worked with OpenAPI definitions).

A quick test to check whether that's the issue would be to change your `RecipeIngredient` definition for this (make sure the JSON format is correct though, I haven't tested it):

```json
{
  "$schema": "http://json-schema.org/schema#",
  "type": "object",
  "required": ["id", "name"],
  "properties": {
    "id": {
      "readOnly": true,
      "type": "integer",
      "example": 1
    },
    "description": {
      "type": "string",
      "example": "Name of the ingredient"
    },
    "name": {
      "type": "string",
      "example": "Sugar"
    }
  }
}
```

And to remove `allOf` so it doesn't get in the way. Then try your call again and see if now both `id` and `name` are required.

Note that in your OpenAPI definition you seem to have declared both `name` and `description` as required, although you told me the required fields should be `id` and `name`. The above assumes the latter is correct (although I don't see where you declare `id` as mandatory in your OpenAPI definition currently).

**Comment by osteel** [comment]
Hmm It's gonna be difficult for me to help without more context I think, but how about you change your YAML definition for this, so we can see if that works:

```yaml
  /recipe/ingredient:

    post:
      tags:
        - recipe-tables
      summary: Add a new recipe ingredient
      description: Add a new recipe ingredient
      operationId: addRecipeIngredient
      requestBody:
        description: Add a new recipe ingredient
        content:
          application/json:
            schema:
              type: object
              required:
                - id
                - name
                - description
              properties:
                id:
                  readOnly: true
                  type: integer
                  example: 1
                description:
                  type: string
                  example: "Name of the ingredient"
                name:
                  type: string
                  example: "Sugar"
      responses:
        '201':
          description: Successful operation
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/RecipeIngredient'
        '400':
          $ref: '#/components/responses/BadRequest'
```

From experience composing schemas from separate files is a bit tricky, and I find that starting from a full YAML definition and then break it down step by step helps.

**Comment by osteel** [comment]
I'm not sure it would return all required missing fields at once – have you tried providing a `name` and no `description`, then?

**Comment by osteel** [comment]
Oh I see! You're using this package in production, as a form validation tool. That's not really what it's intended for though – the main idea is to have a test suite that ensures that your API endpoints don't diverge from your OpenAPI documentation.

As for production environments, you'd typically use a framework with built-in form validation capabilities, and you'd use that instead.

As you may or may not know, this package is built on top of [another one](https://github.com/thephpleague/openapi-psr7-validator), which deals with the actual validation. It does not support returning all errors at once I'm afraid, although there's an [ongoing discussion](https://github.com/thephpleague/openapi-psr7-validator/issues/35) about that.

So I don't think that's gonna work for your use case, although I must say that using OpenAPI definitions for form validation (and thus skip implementing it manually within a framework, for instance) is an interesting idea.
